fix PR 839 changes

jm-c opened the pull request:
## What changes does this PR introduce?
fix 2 bugs in PR #839 changes:

1.  missing indices "i,j" when local scalar vars `chl_sat_sum` and `sat_atten_sum` were changed to local 2-D arrays.
2. second k-loop was merged into first one, the effective irradiance was not set correctly in the mixed-layer with `ML_MEAN_LIGHT` defined.

## What is the current behaviour? 
1. missing indices "(i,j)" when initialising these 2 vars. As a consequence:

   1. compiling with strict F77 compiler fails, like here: 
    http://mitgcm.org/testing/results/2024_06/tr_svante-pgiMth_20240620_0/
   2. when using F90 tolerant compiler, the full 2-D arrays are initialised for each i,j 

2. as second k-loop was removed, with `ML_MEAN_LIGHT` defined, the effective irradiance `irr_eff` is not set correctly in the mixed-layer.

## What is the new behaviour 
fixed

## Does this PR introduce a breaking change? 
With  `ML_MEAN_LIGHT` defined, adding back the second k-loop enable to reproduce ptracers results from before PR #839. But this option was not tested before PR #839 and this fix does not seem to affect AD vars stats in current adjoint test-exp. `global_oce_biogeo_bling` (only forward prtracers are affected).

## Other information:

## Suggested addition to `tag-index`
not needed if merged just after PR #839

jm-c commented:
@mjlosch and @averdy  I pushed this quick fix in `bling_light.F` but I wonder if it would not be better for the 4 new local 2-D arrays that were added in PR #839, i.e.,  `SumMLIrr`, `tmp_ML`, `sat_atten_sum` and `chl_sat_sum`, to initialize them over the full range before starting the k-loop. It would be safer, especially since `tmp_ML` is stored by TAF before beeing initialized.

mjlosch commented:
I think you are right. I didn't think of this and just stupidly/mechanistically added the store directive, even you it stores an uninitialised field (should always ring a bell).

I'll add the initialisation and remove the store directive.

mjlosch commented:
There's another thing that I noticed now with `tmp_ML`. At the bottom of the code, the `irr_eff` is computed like this (as a left-over from combining the loops):
```
#ifdef ML_MEAN_LIGHT
c  Mean irradiance in the mixed layer
          IF ((-rf(k+1) .le. mld(i,j)).and.
     &        (-rf(k+1).lt.MLmix_max)) THEN
           SumMLIrr(i,j) = SumMLIrr(i,j)+drF(k)*irr_inst(i,j,k)
           tmp_ML(i,j) = tmp_ML(i,j) + drF(k)
           irr_mix(i,j) = SumMLIrr(i,j)/tmp_ML(i,j)
          ENDIF
#endif

C Update diagnostic only if ~13:30 local time, when satellite observes
          IF ( localTime.GT.chlsat_locTimWindow(1) .AND.
     &         localTime.LT.chlsat_locTimWindow(2) ) THEN
           chl_sat(i,j,bi,bj) = chl_sat_sum(i,j)/
     &          (sat_atten_sum(i,j) + epsln)
          ENDIF

          irr_eff(i,j,k) = irr_inst(i,j,k)

#ifdef ML_MEAN_LIGHT
c  Inside mixed layer, effective light is set to mean mixed layer light
          IF ((-rf(k+1) .le. mld(i,j)).and.
     &        (-rf(k+1).lt.MLmix_max)) THEN
           irr_eff(i,j,k) = irr_mix(i,j)
          ENDIF
#endif
```
We could simplify this to:
```
#ifdef ML_MEAN_LIGHT
C  Mean irradiance in the mixed layer
          IF ( (-rf(k+1) .LE. mld(i,j)) .AND.
     &         (-rf(k+1) .LT. MLmix_max) ) THEN
           SumMLIrr(i,j)  = SumMLIrr(i,j)+drF(k)*irr_inst(i,j,k)
           tmp_ML(i,j)    = tmp_ML(i,j) + drF(k)
C  Inside mixed layer, effective light is set to mean mixed layer light
           irr_eff(i,j,k) = SumMLIrr(i,j)/tmp_ML(i,j)
          ELSE
#endif
           irr_eff(i,j,k) = irr_inst(i,j,k)
#ifdef ML_MEAN_LIGHT
          ENDIF
#endif

C Update diagnostic only if ~13:30 local time, when satellite observes
          IF ( localTime.GT.chlsat_locTimWindow(1) .AND.
     &         localTime.LT.chlsat_locTimWindow(2) ) THEN
           chl_sat(i,j,bi,bj) = chl_sat_sum(i,j)/
     &          (sat_atten_sum(i,j) + epsln)
          ENDIF
```
thereby separating the diagnostics from the actual computation and saving a 2D field (`irr_mld`). The TAF-AD code would like cleaner, too.

jm-c commented:
@mjlosch I suggest you push the first set of changes (just to have these 4 2-D arrays initialized over full range, outside the k loop) since it seems you already have done this.
And regarding second set of changes, i would like to have @averdy opinion on that before making changes.

averdy commented:
I like the suggestion @mjlosch On Jun 21, 2024, at 11:45 AM, Jean-Michel Campin ***@***.***> wrote:﻿
@mjlosch I suggest you push the first set of changes (just to have these 4 2-D arrays initialized over full range, outside the k loop) since it seems you already have done this.
And regarding second set of changes, i would like to have @averdy opinion on that before making changes.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

jm-c commented:
I took a second look at  changes in PR #839 and have the impression that there are more things to fix ; will try to check soon and add a comment here.

jm-c commented:
I think I restore the correct setting of `irr_eff` in the mixed-layer fro the case `ML_MEAN_LIGHT` defined, as it was before PR #839.
Also, regarding diagnostic field `chl_sat`, I set it to zero when the locat-time is outside the time-window, and this will change the diagnostics output "BLGCHLSA" (previously, when over the `chlsat_locTimWindow(2)`, `chl_sat` was not reset and non-zero value stored in common block from previous time-step was used to fill-in the diagnostics).

This does change the forward ptracers Monitor for the `global_oce_biogeo_bling` AD test but not the AD monitor.

jm-c commented:
And just checked about this:
> This does change the forward ptracers Monitor for the global_oce_biogeo_bling AD test but not the AD monitor.

but theses are identical to before PR #839.

mjlosch commented:
I think that my suggestion of moving the last 3D-loop into the loop before the `C Satellite chlorophyll` in line 393, is still useful and I will push this.

mjlosch commented:
Yes, exactly, we need that, but it doesn't have to be in a separate k-loop, we can use the "first k loop", as far as I can see. The diagnostics (computing `chl_sat`, `chl_sat_sum`, `sat_atten_sum`) can be moved to the end into a separate section.

mjlosch commented on pkg/bling/bling_light.F:
```diff
@@ -381,33 +375,59 @@ SUBROUTINE BLING_LIGHT(
+     &       localTime.LT.chlsat_locTimWindow(2) ) THEN
+         chl_sat(i,j,bi,bj) = chl_sat_sum(i,j)
+     &                      / (sat_atten_sum(i,j) + epsln)
+        ELSE
+         chl_sat(i,j,bi,bj) = 0. _d 0
```
I think that this is not right, because this will feed chl_sat=0 to diagnostics_fill most of the time, thereby "diluting" the average towards zero. The way I am reading the old code (git hash 95edcfd93 and earlier) is that `chl_sat` remains constant if we are outside of this time window. Maybe that's not intentional, but @averdy should check if this is now correct or not.

jm-c replied:
Yes, this does change the diagnostics output "BLGCHLSA" (as mentionned earlier on Jun 22, but will  need @averdy opinion about what is the real purpose of this diagnotics.
Now if we remove these 2 lines (410 + 411) then we also need to add `chl_sat` to the pickup file.

jm-c replied:
Also, the variable `chl_sat` itself is not a time-mean, so this concern
> "diluting" the average towards zero.

does not strictly apply to it but just to the diagnostics  "BLGCHLSA"  when using time-average option (freq > 0). And if this is the only real concern then this could be fixed by adding a "counter-diagnostics" for "BLGCHLSA".

mjlosch replied:
I was referring to the averaging by diagnostics package, so let's add the counter-diagnostics.

I assume that using `chl_sat` in some cost function later will mean that there are no restarts anticipated; something that @averdy needs to decide.

averdy replied:
The main purpose of creating the chl_sat variable is for use with the “generic cost” calculations. Since it’s purely a diagnostic, I don’t think it’s needed in the pickup file. My thought was to update its value only during the mid-day window and keep it constant for the rest of the day, so that the daily average m_chl_sat will be determined by the mid-day value. @mmazloff what are your thoughts?

mmazloff replied:
The passive optical satellites only observe the ocean around local noon, and this value is given to represent the daily measurement. The diagnostic chl_sat approximately mimics this by only updating around local noon. So we don't need a counter or 
        ELSE
         chl_sat(i,j,bi,bj) = 0. _d 0
There are likely better ways to build this observational operator, and we are open to suggestions. However this is working for us now and is useful to other users as well.

jm-c replied:
Thanks for the clarification, it's good when everyone agree. I've  just removed this resetting of "chl_sat" to zero so that, outside the time-window, the value stored in common block from previous time-step will be kept. And no addition to pickup-file for now, knowing that this field will be different as the model restart, at least until it hits the time-window.

jm-c commented:
@averdy I invited you to the "collaborator" team (you will need to accept) so that you could review this PR. Thanks.

jm-c commented:
So , since the problematic changes have been reverted, this PR is opened again for review. @averdy @mmazloff and @mjlosch if you want to approve it I will merge it soon after.

averdy reviewed (commented):
Should “imin,imax” be replaced with “1-OLx,sNx+OLx”?

averdy reviewed (commented):
This line can be removed from under “global variables”: 
C     irr_inst      :: Instantaneous irradiance

jm-c commented:
@averdy
> Should “imin,imax” be replaced with “1-OLx,sNx+OLx”?

When initializing local array, it's better (safer and faster) to do it over the full range. But computation is done only over "i/j min/max" and we can keep it this way as it's done consistently.

And regarding this: 
> This line can be removed from under “global variables”: 
> C irr_inst :: Instantaneous irradiance

I can do this (but this PR did not touch these comments before) and should I move comments related to "irr_mem" elsewhere (e.g., in `BLING_VARS.h`) ? I don't see this variable defined or used anywhere in `bling_light.F`.

averdy commented:
@jm-c Yes it would be great to move to BLING_VARS.h - thank you!

jm-c commented:
@averdy I removed `irr_inst` from the global-variable description/list and moved the description of "irr_mem" to `BLING_VARS.h`. Please let us know if this needs more changes.
Also, I talked to @mjlosch last week and both agree that the second k-loop is needed, as it has been restored in this PR. Will update this PR description to mention that.
So once @averdy you agree/approve this PR changes we should be ready to merge it.

jm-c commented:
I am going to merged this PR soon (later today ?), unless someone wants to add something.
